libvirt_test.go

```diff
@@ -154,6 +154,40 @@ func TestLostConnectionCleanup(t *testing.T) {
 	}
 }
 
+func TestMonitorDisconnect(t *testing.T) {
+	dialer := libvirttest.New()
+	lv := NewWithDialer(dialer)
+
+	// Haven't connected yet.
+	select {
+	case <-lv.Disconnected():
+	default:
+		t.Fatalf("chan is open but should be closed")
+	}
+
+	err := lv.Connect()
+	if err != nil {
+		t.Fatalf("connect failed: %v", err)
+	}
+
+	select {
+	case <-lv.Disconnected():
+		t.Fatalf("chan should be open but is closed")
+	default:
+	}
+
+	err = lv.Disconnect()
+	if err != nil {
+		t.Fatalf("disconnect failed: %v", err)
+	}
+
+	select {
```

Comment:
I worry this case may be racy, is the close of the channel sync in Disconnect()? I think that there is a goro that reacts to socket close and closes the channel, so that it is not guaranteed that the chan is closed _before Disconnect() returns_. It might be better to not have the default in this select, in which case, on failure (the chan never gets closed), the test would eventually timeout and fail, but on success it will be as quick.

Reply:
Disconnect synchronizes with the goro you mentioned (Libvirt.waitAndDisconnect). The goro does some cleanup when it successfully reads from the Libvirt.socket.Disconnected chan and then it closes Libvirt.disconnected which is what Disconnect is synchronizing on before returning.